features/cloud.py

```diff
@@ -462,3 +493,156 @@ def _create_instance(
         return inst
+
+
+class _LXD(Cloud):
+    name = "_lxd"
+
+    def _create_instance(
+        self,
+        series: str,
+        instance_name: "Optional[str]" = None,
+        image_name: "Optional[str]" = None,
+        user_data: "Optional[str]" = None,
+    ) -> pycloudlib.instance:
+        """Launch an instance on the cloud provider.
+
+        :param series:
+            The ubuntu release to be used when creating an instance. We will
+            create an image based on this value if the used does not provide
+            a image_name value
+        :param instance_name:
+            The name of the instance to be created
+        :param image_name:
+            The name of the image to be used when creating the instance
+        :param user_data:
+            The user data to be passed when creating the instance
+
+        :returns:
+            An AWS cloud provider instance
+        """
+        if not image_name:
+            image_name = self.locate_image_name(series)
+
+        image_type = self.name.title().replace("-", " ")
+
+        print(
+            "--- Launching {} image {}({})".format(
+                image_type, image_name, series
+            )
+        )
+
+        inst = self.api.launch(
+            name=instance_name, image_id=image_name, user_data=user_data
+        )
+        return inst
+
+    def get_instance_id(
+        self, instance: pycloudlib.instance.BaseInstance
+    ) -> str:
+        """Return the instance identifier.
+
+        :param instance:
+            An instance created on the cloud provider
+
+        :returns:
+            The string of the unique instance id
+        """
+        # For LXD, the API identifier uses the instance name
+        # instead of the instance id
+        return instance.name
+
+    def locate_image_name(self, series: str) -> str:
+        """Locate and return the image name to use for vm provision.
+
+        :param series:
+            The ubuntu release to be used when locating the image name
+
+        :returns:
+            A image name to use when provisioning a virtual machine
+            based on the series value
+        """
+        if not series:
+            raise ValueError(
+                "Must provide either series or image_name to launch azure"
+            )
+
+        image_name = self.api.daily_image(release=series)
+        return image_name
+
+
+class LXDVirtualMachine(_LXD):
+    name = "lxd-virtual-machine"
+
+    def manage_ssh_key(self, private_key_path: "Optional[str]" = None) -> None:
+        """Create and manage ssh key pairs to be used in the cloud provider.
+
+        :param private_key_path:
+            Location of the private key path to use. If None, the location
+            will be a default location.
+        """
+        pub_key_path = "lxd-pubkey"
+        priv_key_path = "lxd-privkey"
+        pub_key, priv_key = self.api.create_key_pair()
+
+        with open(pub_key_path, "w") as f:
+            f.write(pub_key)
```

Comment:
We probably want to os.chmod these as 600 so that developers could use them on the cmdline to access instances using `ssh -i lxd-pubkey ubuntu@<ip>`